### PR TITLE
feat(role): add valheim server

### PR DESCRIPTION
**Pull request by heihachi**
# Description

Role for creating a simple valheim server using a docker image ([mbround18/valheim](https://github.com/mbround18/valheim-docker)). Files are stored in `/opt/valheim` with saves in `/opt/valheim/saves` and server files in `/opt/valheim/server`.

# How Has This Been Tested?
- [x] Can connect to server
- [x] Continued to run without errors

### CODE REVIEW

File: roles/valheim/defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+##########################################################################
+#                   GNU General Public License v3.0                      #
+##########################################################################
+---
+################################
+# Basics
+################################
+
+valheim_name: valheim
+valheim_version: "LATEST"
```
Comment (saltydk) [inline roles/valheim/defaults/main.yml:15]: Not used for anything.

File: roles/valheim/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+valheim_docker_image_pull: true
+valheim_docker_image_tag: "latest"
+valheim_docker_image: "mbround18/valheim:{{ valheim_docker_image_tag }}"
+
+# Ports
+valheim_docker_ports_defaults:
+  - "2456:2456/udp"
+  - "2457:2457/udp"
+  - "2458:2458/udp"
+valheim_docker_ports: "{{ valheim_docker_ports_defaults }}"
```
Comment (saltydk) [inline roles/valheim/defaults/main.yml:58]: Not following our defaults/custom layout.

File: roles/valheim/defaults/main.yml [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+# Envs
+valheim_docker_envs_default:
+  TZ: "{{ tz }}"
+  UID: "{{ uid }}"
+  GID: "{{ gid }}"
+  PUBLIC: "0"
+  PASSWORD: "{{ lookup('community.general.random_string') }}"
```
Comment (saltydk) [inline roles/valheim/defaults/main.yml:66]: This should be a separate variable so users can override it if they want. Both PUBLIC and PASSWORD, imho.

File: roles/valheim/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  ansible.builtin.include_tasks: "{{ resources_tasks_path }}/docker/remove_docker_container.yml"
+
+- name: Add DNS record
+  ansible.builtin.include_tasks: "{{ resources_tasks_path }}/dns/tasker.yml"
+  vars:
+    dns_record: "{{ lookup('vars', role_name + '_dns_record') }}"
+    dns_zone: "{{ lookup('vars', role_name + '_dns_zone') }}"
+    dns_proxy: "{{ lookup('vars', role_name + '_dns_proxy') }}"
+
+- name: Add SRV record
```
Comment (saltydk) [inline roles/valheim/tasks/main.yml:20]: This whole block seems to be copied from minecraft without ever editing it making this kind of useless. Does this Valheim need the SRV record? If it does then adapt it as needed.

**Review by saltydk — CHANGES_REQUESTED** [review]
@owine or @maximuskowalski can deal with the general formatting, I've commented on the functionality.

File: roles/valheim/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    zone: "{{ _dns_tasker_zone }}"
+    record: "{{ lookup('vars', role_name + '_dns_record') }}"
+    type: SRV
+    proto: udp
+    service: minecraft
+    value: "{{ lookup('vars', role_name + '_dns_record') }}.{{ lookup('vars', role_name + '_dns_zone') }}"
+    port: 19132
+    account_api_token: "{{ cloudflare.api }}"
+    account_email: "{{ cloudflare.email }}"
+  when: cloudflare_is_enabled
```
[the remark this answers is not in the document]
Comment (heihachi) [reply roles/valheim/tasks/main.yml:31]: Whoops missed this as I was using the Minecraft one as a template.

File: roles/valheim/defaults/main.yml (hunk 3 above)
Comment (heihachi) [reply roles/valheim/defaults/main.yml:66]: Could you link an example on how to do this best? I had a variable for the password but when using it in there server name and password env parts they were both different.

**Comment by saltydk** [comment]
I would suggest asking for guidance on the discord if you want help with how to do certain things in a role.

File: roles/valheim/defaults/main.yml (hunk 3 above)
Comment (owine) [reply roles/valheim/defaults/main.yml:66]: The tandoor role requires a secret key that is set in the sandbox settings.yml file. That could be an example for the password. Agree with salty, the public bit could be a simple variable.

**Pull request closed without merging**